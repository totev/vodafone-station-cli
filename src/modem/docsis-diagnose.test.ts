import DocsisDiagnose, { FixWithinOneMonth, checkSignalToNoise, downstreamDeviation, DownstreamDeviation64QAM, downstreamDeviationFactory, FixImmediately, ToleratedDeviation, upstreamDeviation, upstreamDeviationFactory, UpstreamDeviationOFDMA, UpstreamDeviationSCQAM, CompliesToSpecifications } from "./docsis-diagnose";
import type { DocsisChannelType, DocsisStatus, Modulation } from "./modem";
import fixtureDocsisStatusArris from './__fixtures__/docsisStatus_arris_normalized.json';
import fixtureDocsisStatus from './__fixtures__/docsisStatus_normalized.json';
import fixtureDocsisStatusMinimal from './__fixtures__/docsisStatus_normalized_minimal.json';

test('constructor', () => {
  const diagnoser = new DocsisDiagnose(fixtureDocsisStatus as DocsisStatus);
  expect(diagnoser).toBeTruthy();
});

test('checkDownstream', () => {
  const diagnoser = new DocsisDiagnose(fixtureDocsisStatus as DocsisStatus);
  const result = diagnoser.checkDownstream() 
  expect(result).toHaveLength(32);
  expect(result).toMatchSnapshot();
});

test('checkOfdmDownstream', () => {
  const diagnoser = new DocsisDiagnose(fixtureDocsisStatus as DocsisStatus);
  const result = diagnoser.checkOfdmDownstream() 
  expect(result).toHaveLength(1);
  expect(result).toMatchSnapshot();
});

test('checkUpstream', () => {
  const diagnoser = new DocsisDiagnose(fixtureDocsisStatus as DocsisStatus);
  const result = diagnoser.checkUpstream() 
  expect(result).toHaveLength(4);
  expect(result).toMatchSnapshot();
});

test('checkOfdmaUpstream', () => {
  const diagnoser = new DocsisDiagnose(fixtureDocsisStatus as DocsisStatus);
  const result = diagnoser.checkOfdmaUpstream() 
  expect(result).toHaveLength(1);
  expect(result).toMatchSnapshot();
});

test('deviationFactory unsupported modulation/type', () => {
  expect(()=>downstreamDeviationFactory("8192AM" as Modulation)).toThrow()
  expect(()=>upstreamDeviationFactory("OFDM" as DocsisChannelType)).toThrow()
});
test('deviationFactory supported modulation/type', () => {
  expect(downstreamDeviationFactory("64QAM" as Modulation)).toStrictEqual(new DownstreamDeviation64QAM())
  expect(upstreamDeviationFactory("SC-QAM" as const)).toStrictEqual(new UpstreamDeviationSCQAM())
  expect(upstreamDeviationFactory("OFDMA" as const)).toStrictEqual(new UpstreamDeviationOFDMA())
});

test('hasDeviations with deviations', () => {
  const diagnoser = new DocsisDiagnose(fixtureDocsisStatus as DocsisStatus);
  expect(diagnoser.hasDeviations()).toBeTruthy();
});

test('hasDeviations without deviations', () => {
  const diagnoser = new DocsisDiagnose(fixtureDocsisStatusMinimal as DocsisStatus);
  expect(diagnoser.hasDeviations()).toBeFalsy();
});


describe('diagnoseUpstream', () => {
  test.each([
    [-70,	 FixImmediately],
    [-56,	 FixImmediately],
    [-35,	 FixImmediately],
    [-4,	 FixImmediately],
    [0,	 FixImmediately],
    [10,	 FixImmediately],
    [35,	 FixImmediately],
    [35.1, FixWithinOneMonth],
    [36, FixWithinOneMonth],
    [37, FixWithinOneMonth],
    [37.1, ToleratedDeviation],
    [40, ToleratedDeviation],
    [41, ToleratedDeviation],
    [41.1, CompliesToSpecifications],
    [45, CompliesToSpecifications],
    [47, CompliesToSpecifications],
    [47.1, ToleratedDeviation],
    [49, ToleratedDeviation],
    [51, ToleratedDeviation],
    [51.1, FixWithinOneMonth],
    [52, FixWithinOneMonth],
    [53, FixWithinOneMonth],
    [53.1,	 FixImmediately],
    [60,	 FixImmediately],
    [70,	 FixImmediately],
  ])
  ('upstreamDeviationDeviationSCQAM(%d)', (input, expected ) => {
    expect(upstreamDeviation({powerLevel:input, channelType:"SC-QAM"})).toBe(expected);
  });
  
  test.each([
    [-60,	 FixImmediately],
    [-50,	 FixImmediately],
    [-30,	 FixImmediately],
    [-4,	 FixImmediately],
    [0,	 FixImmediately],
    [10,	 FixImmediately],
    [20,	 FixImmediately],
    [30,	 FixImmediately],
    [38,	 FixImmediately],
    [38.1, FixWithinOneMonth],
    [39, FixWithinOneMonth],
    [39.9, FixWithinOneMonth],
    [40, FixWithinOneMonth],
    [40.1, ToleratedDeviation],
    [41, ToleratedDeviation],
    [42, ToleratedDeviation],
    [44, ToleratedDeviation],
    [44.1, CompliesToSpecifications],
    [45, CompliesToSpecifications],
    [47, CompliesToSpecifications],
    [47.1, ToleratedDeviation],
    [47.55, ToleratedDeviation],
    [48, ToleratedDeviation],
    [48.1, FixWithinOneMonth],
    [49, FixWithinOneMonth],
    [50, FixWithinOneMonth],
    [50.1,	 FixImmediately],
    [60,	 FixImmediately],
    [70,	 FixImmediately],
  ])
  ('upstreamDeviationDeviationOFDMA(%d)', (input, expected ) => {
    expect(upstreamDeviation({powerLevel:input, channelType:"OFDMA"})).toBe(expected);
  });
});

describe('diagnoseDownstream', () => {

  test.each`
    input	| expected
    ${{modulation:"64QAM", powerLevel:-60}}	| ${FixImmediately}
    ${{modulation:"64QAM", powerLevel:-50}}	| ${FixImmediately}
    ${{modulation:"64QAM", powerLevel:-14}}	| ${FixImmediately}
    ${{modulation:"64QAM", powerLevel:-13.9}}	| ${FixWithinOneMonth}
    ${{modulation:"64QAM", powerLevel:-13}}	| ${FixWithinOneMonth}
    ${{modulation:"64QAM", powerLevel:-12}}	| ${FixWithinOneMonth}
    ${{modulation:"64QAM", powerLevel:-11.9}}	| ${ToleratedDeviation}
    ${{modulation:"64QAM", powerLevel:-10.5}}	| ${ToleratedDeviation}
    ${{modulation:"64QAM", powerLevel:-10}}	| ${ToleratedDeviation}
    ${{modulation:"64QAM", powerLevel:-9.9}}	| ${CompliesToSpecifications}
    ${{modulation:"64QAM", powerLevel:-9}}	| ${CompliesToSpecifications}
    ${{modulation:"64QAM", powerLevel:-8}}	| ${CompliesToSpecifications}
    ${{modulation:"64QAM", powerLevel:-7}}	| ${CompliesToSpecifications}
    ${{modulation:"64QAM", powerLevel:0}}	| ${CompliesToSpecifications}
    ${{modulation:"64QAM", powerLevel:5}}	| ${CompliesToSpecifications}
    ${{modulation:"64QAM", powerLevel:7}}	| ${CompliesToSpecifications}
    ${{modulation:"64QAM", powerLevel:7.1}}	| ${ToleratedDeviation}
    ${{modulation:"64QAM", powerLevel:10}}	| ${ToleratedDeviation}
    ${{modulation:"64QAM", powerLevel:11.99}}	| ${ToleratedDeviation}
    ${{modulation:"64QAM", powerLevel:12}}	| ${ToleratedDeviation}
    ${{modulation:"64QAM", powerLevel:12.1}}	| ${FixWithinOneMonth}
    ${{modulation:"64QAM", powerLevel:13}}	| ${FixWithinOneMonth}
    ${{modulation:"64QAM", powerLevel:14}}	| ${FixWithinOneMonth}
    ${{modulation:"64QAM", powerLevel:14.1}}	| ${FixImmediately}
    ${{modulation:"64QAM", powerLevel:100}}	| ${FixImmediately}
  `('downstreamDeviation($input)', ({ input, expected }) => {
    expect(downstreamDeviation(input)).toBe(expected);
  });

  test.each`
    input	| expected
    ${{modulation:"256QAM", powerLevel:-60}}	| ${FixImmediately}
    ${{modulation:"256QAM", powerLevel:-50}}	| ${FixImmediately}
    ${{modulation:"256QAM", powerLevel:-8}}	| ${FixImmediately}
    ${{modulation:"256QAM", powerLevel:-7.9}}	| ${FixWithinOneMonth}
    ${{modulation:"256QAM", powerLevel:-7}}	| ${FixWithinOneMonth}
    ${{modulation:"256QAM", powerLevel:-6}}	| ${FixWithinOneMonth}
    ${{modulation:"256QAM", powerLevel:-5.9}}	| ${ToleratedDeviation}
    ${{modulation:"256QAM", powerLevel:-4.5}}	| ${ToleratedDeviation}
    ${{modulation:"256QAM", powerLevel:-4}}	| ${ToleratedDeviation}
    ${{modulation:"256QAM", powerLevel:-3.9}}	| ${CompliesToSpecifications}
    ${{modulation:"256QAM", powerLevel:-3}}	| ${CompliesToSpecifications}
    ${{modulation:"256QAM", powerLevel:-2}}	| ${CompliesToSpecifications}
    ${{modulation:"256QAM", powerLevel:-1}}	| ${CompliesToSpecifications}
    ${{modulation:"256QAM", powerLevel:0}}	| ${CompliesToSpecifications}
    ${{modulation:"256QAM", powerLevel:11}}	| ${CompliesToSpecifications}
    ${{modulation:"256QAM", powerLevel:13}}	| ${CompliesToSpecifications}
    ${{modulation:"256QAM", powerLevel:13.1}}	| ${ToleratedDeviation}
    ${{modulation:"256QAM", powerLevel:16}}	| ${ToleratedDeviation}
    ${{modulation:"256QAM", powerLevel:17.99}}	| ${ToleratedDeviation}
    ${{modulation:"256QAM", powerLevel:18}}	| ${ToleratedDeviation}
    ${{modulation:"256QAM", powerLevel:18.1}}	| ${FixWithinOneMonth}
    ${{modulation:"256QAM", powerLevel:19}}	| ${FixWithinOneMonth}
    ${{modulation:"256QAM", powerLevel:20}}	| ${FixWithinOneMonth}
    ${{modulation:"256QAM", powerLevel:20.1}}	| ${FixImmediately}
    ${{modulation:"256QAM", powerLevel:100}}	| ${FixImmediately}
  `('downstreamDeviation($input)', ({ input, expected }) => {
    expect(downstreamDeviation(input)).toBe(expected);
  });

  
  test.each([
    [-60,	 FixImmediately],
    [-50,	 FixImmediately],
    [-6,	 FixImmediately],
    [-5.9, FixWithinOneMonth],
    [-5,	 FixWithinOneMonth],
    [-4,	 FixWithinOneMonth],
    [-3.9, ToleratedDeviation],
    [-2.5, ToleratedDeviation],
    [-2,	 ToleratedDeviation],
    [-1.9,	 CompliesToSpecifications],
    [-1,	   CompliesToSpecifications],
    [1,	   CompliesToSpecifications],
    [0,	   CompliesToSpecifications],
    [5,	   CompliesToSpecifications],
    [15,	   CompliesToSpecifications],
    [15.1,	 ToleratedDeviation],
    [18,	   ToleratedDeviation],
    [19.99, ToleratedDeviation],
    [20,	   ToleratedDeviation],
    [20.1,	 FixWithinOneMonth],
    [21,	   FixWithinOneMonth],
    [22,	   FixWithinOneMonth],
    [22.1,	 FixImmediately],
    [100,	 FixImmediately],
  ])
  ('downstreamDeviation1024QAM(%d)', (input, expected ) => {
    expect(downstreamDeviation({powerLevel:input, modulation:"1024QAM"})).toBe(expected);
  });
  
  test.each([
    [-56,	 FixImmediately],
    [-48,	 FixImmediately],
    [-4,	 FixImmediately],
    [-3.9, FixWithinOneMonth],
    [-3,	 FixWithinOneMonth],
    [-2,	 FixWithinOneMonth],
    [-1.9, ToleratedDeviation],
    [-0.5, ToleratedDeviation],
    [0,	 ToleratedDeviation],
    [0.1,	 CompliesToSpecifications],
    [1,	   CompliesToSpecifications],
    [7,	   CompliesToSpecifications],
    [8,	   CompliesToSpecifications],
    [13,	   CompliesToSpecifications],
    [17,	   CompliesToSpecifications],
    [17.1,	 ToleratedDeviation],
    [20,	   ToleratedDeviation],
    [21.99, ToleratedDeviation],
    [22,	   ToleratedDeviation],
    [22.1,	 FixWithinOneMonth],
    [23,	   FixWithinOneMonth],
    [24,	   FixWithinOneMonth],
    [24.1,	 FixImmediately],
    [102,	 FixImmediately],
  ])
  ('downstreamDeviation2048QAM(%d)', (input, expected ) => {
    expect(downstreamDeviation({powerLevel:input, modulation:"2048QAM"})).toBe(expected);
  });


  test.each([
    [-54,	 FixImmediately],
    [-46,	 FixImmediately],
    [-2,	 FixImmediately],
    [-1.9, FixWithinOneMonth],
    [-1,	 FixWithinOneMonth],
    [0,	 FixWithinOneMonth],
    [0.1, ToleratedDeviation],
    [1.5, ToleratedDeviation],
    [2,	 ToleratedDeviation],
    [2.1,	 CompliesToSpecifications],
    [3,	   CompliesToSpecifications],
    [9,	   CompliesToSpecifications],
    [10,	   CompliesToSpecifications],
    [15,	   CompliesToSpecifications],
    [19,	   CompliesToSpecifications],
    [19.1,	 ToleratedDeviation],
    [22,	   ToleratedDeviation],
    [23.99, ToleratedDeviation],
    [24,	   ToleratedDeviation],
    [24.1,	 FixWithinOneMonth],
    [25,	   FixWithinOneMonth],
    [26,	   FixWithinOneMonth],
    [26.1,	 FixImmediately],
    [104,	 FixImmediately],
  ])
  ('downstreamDeviation4096QAM(%d)', (input, expected ) => {
    expect(downstreamDeviation({powerLevel:input, modulation:"4096QAM"})).toBe(expected);
  });
  
});

describe('signalToNoise', () => {

  test.each([
    [-50,	 FixImmediately],
    [-10,	 FixImmediately],
    [-1,	 FixImmediately],
    [0,	 FixImmediately],
    [10,	 FixImmediately],
    [20,	 FixImmediately],
    [24,	 FixImmediately],
    [24.1, FixWithinOneMonth],
    [25, FixWithinOneMonth],
    [26, FixWithinOneMonth],
    [26.1, ToleratedDeviation],
    [26.6, ToleratedDeviation],
    [27,	 ToleratedDeviation],
    [27.1,	 CompliesToSpecifications],
    [28,	   CompliesToSpecifications],
    [30,	   CompliesToSpecifications],
    [35,	   CompliesToSpecifications],
  ])
  ('checkSignalToNoise64QAM(%d)', (input, expected ) => {
    expect(checkSignalToNoise({ snr: input, modulation: "64QAM" })).toBe(expected);
  });

  test.each([
    [-50,	 FixImmediately],
    [-10,	 FixImmediately],
    [-1,	 FixImmediately],
    [0,	 FixImmediately],
    [10,	 FixImmediately],
    [20,	 FixImmediately],
    [24,	 FixImmediately],
    [29.9,	 FixImmediately],
    [30,	 FixImmediately],
    [30.1, FixWithinOneMonth],
    [31, FixWithinOneMonth],
    [32, FixWithinOneMonth],
    [32.1, ToleratedDeviation],
    [32.6, ToleratedDeviation],
    [33,	 ToleratedDeviation],
    [33.1,	 CompliesToSpecifications],
    [34,	   CompliesToSpecifications],
    [40,	   CompliesToSpecifications],
    [45,	   CompliesToSpecifications],
  ])
  ('checkSignalToNoise256QAM(%d)', (input, expected ) => {
    expect(checkSignalToNoise({ snr: input, modulation: "256QAM" })).toBe(expected);
  });

  test.each([
    [-50,	 FixImmediately],
    [-10,	 FixImmediately],
    [-1,	 FixImmediately],
    [0,	 FixImmediately],
    [10,	 FixImmediately],
    [20,	 FixImmediately],
    [24,	 FixImmediately],
    [35.9,	 FixImmediately],
    [36,	 FixImmediately],
    [36.1, FixWithinOneMonth],
    [37.99, FixWithinOneMonth],
    [38, FixWithinOneMonth],
    [38.1, ToleratedDeviation],
    [38.6, ToleratedDeviation],
    [39,	 ToleratedDeviation],
    [39.1,	 CompliesToSpecifications],
    [40,	   CompliesToSpecifications],
    [47,	   CompliesToSpecifications],
    [50,	   CompliesToSpecifications],
  ])
  ('checkSignalToNoise1024QAM(%d)', (input, expected ) => {
    expect(checkSignalToNoise({ snr: input, modulation: "1024QAM" })).toBe(expected);
  });

  test.each([
    [-50,	 FixImmediately],
    [-10,	 FixImmediately],
    [-1,	 FixImmediately],
    [0,	 FixImmediately],
    [10,	 FixImmediately],
    [20,	 FixImmediately],
    [24,	 FixImmediately],
    [38.9,	 FixImmediately],
    [39,	 FixImmediately],
    [39.1, FixWithinOneMonth],
    [39.99, FixWithinOneMonth],
    [41, FixWithinOneMonth],
    [41.1, ToleratedDeviation],
    [41.6, ToleratedDeviation],
    [42,	 ToleratedDeviation],
    [43.1,	 CompliesToSpecifications],
    [45,	   CompliesToSpecifications],
    [47,	   CompliesToSpecifications],
    [50,	   CompliesToSpecifications],
  ])
  ('checkSignalToNoise2048QAM(%d)', (input, expected) => {
    expect(checkSignalToNoise({ snr: input, modulation: "2048QAM" })).toBe(expected);
  });

  test.each([
    [-50,	 FixImmediately],
    [-10,	 FixImmediately],
    [-1,	 FixImmediately],
    [0,	 FixImmediately],
    [10,	 FixImmediately],
    [20,	 FixImmediately],
    [30,	 FixImmediately],
    [40,	 FixImmediately],
    [41.9,	 FixImmediately],
    [42,	 FixImmediately],
    [42.1, FixWithinOneMonth],
    [42.99, FixWithinOneMonth],
    [43, FixWithinOneMonth],
    [44, FixWithinOneMonth],
    [44.1, ToleratedDeviation],
    [44.6, ToleratedDeviation],
    [45,	 ToleratedDeviation],
    [45.1,	 CompliesToSpecifications],
    [46,	   CompliesToSpecifications],
    [47,	   CompliesToSpecifications],
    [50,	   CompliesToSpecifications],
  ])
  ('checkSignalToNoise4096QAM(%d)', (input, expected ) => {
    expect(checkSignalToNoise({ snr: input, modulation: "4096QAM"})).toBe(expected);
  });

});

test('printDeviationsConsole with pl deviations', () => {
  const diagnoser = new DocsisDiagnose(fixtureDocsisStatusArris as DocsisStatus);
  const result = diagnoser.printDeviationsConsole()
  console.log(result);
  expect(result).toContain("ch33pl");
});

test('printDeviationsConsole without deviations', () => {
  const diagnoser = new DocsisDiagnose(fixtureDocsisStatusMinimal as DocsisStatus);
  const result = diagnoser.printDeviationsConsole()
  console.log(result);
  expect(result).toContain("Hooray");
});

test('printDeviationsConsole with snr deviations', () => {
  const fixture = {
    "downstream": [
      {
        "channelId": "2",
        "channelType": "SC-QAM",
        "modulation": "256QAM",
        "powerLevel": -5.6,
        "lockStatus": "Locked",
        "snr": 31,
        "frequency": 130
      },
    ],
    downstreamOfdm: [],
    upstream:[],
    upstreamOfdma:[],
  } as any as DocsisStatus
  const diagnoser = new DocsisDiagnose(fixture);
  const result = diagnoser.printDeviationsConsole()
  console.log(result);
  expect(result).toContain("snr");
});
