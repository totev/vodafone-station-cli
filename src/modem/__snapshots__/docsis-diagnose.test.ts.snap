// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkDownstream 1`] = `
[
  {
    "channelId": "1",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 114,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 55.1,
    "snr": 35,
  },
  {
    "channelId": "2",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 130,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 54.7,
    "snr": 35,
  },
  {
    "channelId": "3",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 138,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 54.8,
    "snr": 36,
  },
  {
    "channelId": "4",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 146,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 54.6,
    "snr": 36,
  },
  {
    "channelId": "5",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 602,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 57,
    "snr": 38,
  },
  {
    "channelId": "6",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 618,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 57.3,
    "snr": 39,
  },
  {
    "channelId": "7",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 626,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 57.7,
    "snr": 38,
  },
  {
    "channelId": "8",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 642,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 58.5,
    "snr": 39,
  },
  {
    "channelId": "9",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 650,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 58.3,
    "snr": 39,
  },
  {
    "channelId": "10",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 658,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 58.3,
    "snr": 39,
  },
  {
    "channelId": "11",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 666,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 58.1,
    "snr": 38,
  },
  {
    "channelId": "12",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 674,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 58.8,
    "snr": 39,
  },
  {
    "channelId": "13",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 682,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 58.8,
    "snr": 39,
  },
  {
    "channelId": "14",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 690,
    "lockStatus": "Locked",
    "modulation": "256QAM",
    "powerLevel": 59.4,
    "snr": 39,
  },
  {
    "channelId": "15",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 698,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 53,
    "snr": 35,
  },
  {
    "channelId": "16",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 706,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.1,
    "snr": 35,
  },
  {
    "channelId": "17",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 714,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.2,
    "snr": 34,
  },
  {
    "channelId": "18",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 722,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 53.8,
    "snr": 35,
  },
  {
    "channelId": "19",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 730,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 53.9,
    "snr": 35,
  },
  {
    "channelId": "20",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 738,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.9,
    "snr": 35,
  },
  {
    "channelId": "21",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 746,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 55.3,
    "snr": 35,
  },
  {
    "channelId": "22",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 754,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.5,
    "snr": 35,
  },
  {
    "channelId": "23",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 762,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.5,
    "snr": 35,
  },
  {
    "channelId": "24",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 770,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.5,
    "snr": 35,
  },
  {
    "channelId": "25",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 778,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 55,
    "snr": 35,
  },
  {
    "channelId": "26",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 786,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 55,
    "snr": 35,
  },
  {
    "channelId": "27",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 794,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.9,
    "snr": 34,
  },
  {
    "channelId": "28",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 802,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.4,
    "snr": 35,
  },
  {
    "channelId": "29",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 810,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.1,
    "snr": 35,
  },
  {
    "channelId": "30",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 818,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.5,
    "snr": 35,
  },
  {
    "channelId": "31",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 826,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 54.5,
    "snr": 35,
  },
  {
    "channelId": "32",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 834,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 53.7,
    "snr": 34,
  },
]
`;

exports[`checkOfdmDownstream 1`] = `
[
  {
    "channelId": "33",
    "channelType": "OFDM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequencyEnd": 324,
    "frequencyStart": 151,
    "lockStatus": "Locked",
    "modulation": "1024QAM",
    "powerLevel": 56.1,
    "snr": 39,
  },
]
`;

exports[`checkOfdmaUpstream 1`] = `
[
  {
    "channelId": "9",
    "channelType": "OFDMA",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequencyEnd": 64.8,
    "frequencyStart": 29.8,
    "lockStatus": "SUCCESS",
    "modulation": "16_QAM",
    "powerLevel": 106.2,
    "snr": 0,
  },
]
`;

exports[`checkUpstream 1`] = `
[
  {
    "channelId": "3",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 37,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 110.3,
    "snr": 0,
  },
  {
    "channelId": "4",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 31,
    "lockStatus": "Locked",
    "modulation": "32QAM",
    "powerLevel": 110.3,
    "snr": 0,
  },
  {
    "channelId": "1",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 51,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 110.3,
    "snr": 0,
  },
  {
    "channelId": "2",
    "channelType": "SC-QAM",
    "diagnose": {
      "color": "red",
      "description": "Fix immediately",
      "deviation": true,
    },
    "frequency": 45,
    "lockStatus": "Locked",
    "modulation": "64QAM",
    "powerLevel": 110.3,
    "snr": 0,
  },
]
`;
